File: app/src/main/java/com/example/noev/greeter/MainActivity.java
package com.example.noev.greeter;

import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.SeekBar;
import android.widget.TextView;
import android.widget.Toast;

import org.web3j.abi.datatypes.Utf8String;
import org.web3j.crypto.Credentials;
import org.web3j.protocol.Web3j;
import org.web3j.protocol.Web3jFactory;
import org.web3j.protocol.core.methods.response.TransactionReceipt;
import org.web3j.protocol.infura.InfuraHttpService;

import java.math.BigInteger;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;

/**
 * Main Ethereum Network
 * https://mainnet.infura.io/[your-token]
 * <p>
 * Test Ethereum Network (Ropsten)
 * https://ropsten.infura.io/[your-token]
 * <p>
 * Test Rinkeby Network
 * https://rinkeby.infura.io/[your-token]
 * <p>
 * IPFS Gateway
 * https://ipfs.infura.io
 * <p>
 * IPFS RPC
 * https://ipfs.infura.io:5001
 */
public class MainActivity extends AppCompatActivity {

    private final static String privateKeyRopsten = "YOUR_PRIVATE_KEY";
    private final static String greeterContractAddressRopsten = "0x024b64940518779068e57352F3bDDdE08E4D9c40";
    private final static String ropstenUrl = "https://ropsten.infura.io/YOUR_API_KEY";

    private ProgressBar progressBar;
    private EditText editText;
    private TextView greetingTextView;
    private TextView gasPriceTextView;
    private TextView gasLimitTextView;
    private SeekBar gasPriceSeekBar;

    private Web3j web3j;

    private Credentials credentials = Credentials.create(privateKeyRopsten);
    private int minimumGasLimit = 21000;
    private BigInteger gasLimit = new BigInteger(String.valueOf(minimumGasLimit));

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        initUi();
        setGasPriceText(10);
        setGasLimit(minimumGasLimit);
        initWeb3j();
    }

    private void initUi() {
        progressBar = (ProgressBar) findViewById(R.id.progressbar);
        editText = (EditText) findViewById(R.id.edit_text);
        greetingTextView = (TextView) findViewById(R.id.text);
        Button readButton = (Button) findViewById(R.id.button);
        readButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                getGreeting();
            }
        });
        Button writeButton = (Button) findViewById(R.id.write_button);
        writeButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                writeGreetingToContract();
            }
        });
        gasPriceSeekBar = (SeekBar) findViewById(R.id.gas_price_seek_bar);
        gasPriceSeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                setGasPriceText(progress);
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });
        SeekBar gasLimitSeekBar = (SeekBar) findViewById(R.id.gas_limit_seek_bar);
        gasLimitSeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                setGasLimit(progress + minimumGasLimit);
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });
        gasLimitTextView = (TextView) findViewById(R.id.gas_limit_text_view);
        gasPriceTextView = (TextView) findViewById(R.id.gas_price_text_view);
    }

    private void writeGreetingToContract() {
        progressBar.setVisibility(View.VISIBLE);
        WriteTask writeTask = new WriteTask();
        writeTask.execute(editText.getText().toString());
    }

    private void getGreeting() {
        try {
            progressBar.setVisibility(View.VISIBLE);
            ReadTask readTask = new ReadTask();
            readTask.execute();
        } catch (Exception e) {
            Log.d("wat", "getGreeting exception = " + e.getMessage());
        }
    }

    private void initWeb3j() {
        InitWeb3JTask task = new InitWeb3JTask();
        task.execute(ropstenUrl);
    }

    public void setGasPriceText(int gasPrice) {
        String formattedString = getString(R.string.gas_price, String.valueOf(gasPrice));
        gasPriceTextView.setText(formattedString);
    }

    private BigInteger getGasPrice() {
        int gasPriceGwei = gasPriceSeekBar.getProgress();
        BigInteger gasPriceWei = BigInteger.valueOf(gasPriceGwei + 1000000000L);
        Log.d("wat", "getGasPrice: " + String.valueOf(gasPriceGwei));
        return gasPriceWei;
    }

    public void setGasLimit(int gasLimit) {
        String gl = String.valueOf(gasLimit);
        this.gasLimit = new BigInteger(gl);
        gasLimitTextView.setText(getString(R.string.gas_limit, gl));
    }

    private class ReadTask extends AsyncTask<String, String, String> {
        @Override
        protected String doInBackground(String... params) {
            String result;
            try {
                Greeter greeter = Greeter.load(greeterContractAddressRopsten, web3j, credentials, getGasPrice(), gasLimit);
                Future<Utf8String> greeting = greeter.greet();
                Utf8String greetingUtf8 = greeting.get();
                result = greetingUtf8.getValue();
            } catch (Exception e) {
                result = "Error reading the smart contract. Error: " + e.getMessage();
            }

            return result;
        }

        @Override
        protected void onPostExecute(String result) {
            super.onPostExecute(result);
            progressBar.setVisibility(View.INVISIBLE);
            greetingTextView.setText(result);
        }
    }


    private class WriteTask extends AsyncTask<String, String, String> {
        @Override
        protected String doInBackground(String... params) {
            String greetingToWrite = params[0];

            String result;
            try {
                Greeter greeter = Greeter.load(greeterContractAddressRopsten, web3j, credentials, getGasPrice(), gasLimit);
                TransactionReceipt transactionReceipt = greeter.changeGreeting(new Utf8String(greetingToWrite)).get(3, TimeUnit.MINUTES);
                result = "Successful transaction. Gas used: " + transactionReceipt.getGasUsed();
            } catch (Exception e) {
                result = "Error during transaction. Error: " + e.getMessage();
            }

            return result;
        }

        @Override
        protected void onPostExecute(String result) {
            super.onPostExecute(result);
            progressBar.setVisibility(View.INVISIBLE);
            Toast.makeText(MainActivity.this, result, Toast.LENGTH_LONG).show();
        }
    }

    private class InitWeb3JTask extends AsyncTask<String, String, String> {
        @Override
        protected String doInBackground(String... params) {
            String url = params[0];
            InfuraHttpService infuraHttpService;
            String result = "Success initializing web3j/infura";
            try {
                infuraHttpService = new InfuraHttpService(url);
                web3j = Web3jFactory.build(infuraHttpService);
            } catch (Exception wtf) {
                String exception = wtf.toString();
                Log.d("wat", "Error initializing web3j/infura. Error: " + exception);
            }

            return result;
        }

        @Override
        protected void onPostExecute(String result) {
            super.onPostExecute(result);
            Toast.makeText(MainActivity.this, result, Toast.LENGTH_LONG).show();
        }
    }
}

File: app/src/main/java/com/noev/ethereumhelloworld/MainActivity.kt
package com.noev.ethereumhelloworld

import android.databinding.*
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.widget.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.web3j.crypto.Credentials
import org.web3j.protocol.Web3j
import org.web3j.protocol.http.HttpService
import java.math.BigInteger
import java.util.concurrent.TimeUnit

const val MINIMUM_GAS_LIMIT = 21000
const val PRIVATE_KEY_ROPSTEN = "YOUR_ROPSTEN_PRIVATE_KEY" //todo: You have to create an ethereum account on the Ropsten network and put your private key here
const val ROPSTEN_INFURA_URL = "https://ropsten.infura.io/YOUR_API_KEY" //todo: You have to register on the Infura website and put your api key here
const val CONTRACT_ADDRESS = "0x024b64940518779068e57352F3bDDdE08E4D9c40"

class MainActivity : AppCompatActivity() {

    val isLoading = ObservableBoolean()
    val textReadFromContract = ObservableField<String>()
    val gasPrice = ObservableInt(10)
    val userText = ObservableField<String>()
    val gasLimit = ObservableInt(MINIMUM_GAS_LIMIT)

    private var web3j: Web3j? = null

    private val credentials = Credentials.create(PRIVATE_KEY_ROPSTEN)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val binding = DataBindingUtil.setContentView<ViewDataBinding>(this, R.layout.activity_main)
        binding.setVariable(BR.viewModel, this)

        CoroutineScope(Dispatchers.Main).launch(Dispatchers.IO) {
            isLoading.set(true)
            val result = initializeWeb3J()
            isLoading.set(false)
            toast(result)
        }
    }

    fun writeButtonClicked() {
        CoroutineScope(Dispatchers.Main).launch(Dispatchers.IO) {
            isLoading.set(true)
            val result = writeToContract()
            isLoading.set(false)
            toast(result)
        }
    }

    fun readButtonClicked() {
        CoroutineScope(Dispatchers.Main).launch(Dispatchers.IO) {
            isLoading.set(true)
            val result = readFromContract()
            isLoading.set(false)
            textReadFromContract.set(result)
        }
    }

    private fun writeToContract(): String {
        val greetingToWrite = userText.get()
        val result: String
        result = try {
            val greeter = Greeter.load(CONTRACT_ADDRESS, web3j, credentials, getGasPrice(), getGasLimit())
            val transactionReceipt = greeter.changeGreeting(greetingToWrite).sendAsync().get(3, TimeUnit.MINUTES)
            "Successful transaction. Gas used: " + transactionReceipt.gasUsed
        } catch (e: Exception) {
            "Error during transaction. Error: " + e.message
        }
        return result
    }

    private fun readFromContract(): String {
        val result: String
        result = try {
            val greeter = Greeter.load(CONTRACT_ADDRESS, web3j, credentials, getGasPrice(), getGasLimit())
            val greeting = greeter.greet().sendAsync()
            greeting.get()
        } catch (e: Exception) {
            "Error reading the smart contract. Error: " + e.message
        }
        return result
    }

    private fun initializeWeb3J(): String {
        val infuraHttpService: HttpService
        val result: String
        result = try {
            infuraHttpService = HttpService(ROPSTEN_INFURA_URL)
            web3j = Web3j.build(infuraHttpService)
            "Success initializing web3j/infura"
        } catch (e: Exception) {
            val exception = e.toString()
            "Error initializing web3j/infura. Error: $exception"
        }

        return result
    }

    private fun getGasPrice(): BigInteger {
        val gasPriceGwei = gasPrice.get()
        val gasPriceWei = BigInteger.valueOf(gasPriceGwei + 1000000000L)
        Log.d("wat", "getGasPrice: $gasPriceGwei")
        return gasPriceWei
    }

    private fun getGasLimit(): BigInteger {
        return gasLimit.get().bigInteger()
    }

    private fun Int.bigInteger(): BigInteger {
        return BigInteger(toString())
    }

    private fun toast(text: String) {
        runOnUiThread {
            Toast.makeText(this, text, Toast.LENGTH_LONG).show()
        }
    }
}
